File: erpnext_payment_uploading/erpnext_payment_uploading/page/payment_upload/payment_upload.js
frappe.pages["payment-upload"].on_page_load = (wrapper) => {
	const page = frappe.ui.make_app_page({ parent: wrapper, title: __("Cheque Journal Entry Upload"), single_column: true });
	new PaymentUpload(page);
};

class PaymentUpload {
	constructor(page) {
		this.page = page;
		this.rows = [];
		this.debit_rows = [];
		this.make();
	}

	make() {
		this.$body = $(`<div class="payment-upload">
			<div class="frappe-card p-4 mb-4"><div class="row"></div></div>
			<div class="summary mb-3"></div><div class="preview"></div>
			<div class="debit-section mt-4"></div>
		</div>`).appendTo(this.page.main);
		const $row = this.$body.find(".row");
		this.company = this.control($row, "Link", "company", __("Company"), true, "Company");
		this.posting_date = this.control($row, "Date", "posting_date", __("Posting Date"), true);
		this.posting_date.set_value(frappe.datetime.get_today());
		this.ewt_account = this.control($row, "Link", "ewt_account", __("EWT Account"), true, "Account", {
			get_query: () => ({ filters: { company: this.company.get_value(), is_group: 0 } }),
			onchange: () => this.update_balance_state(),
		});
		this.write_off = this.control($row, "Link", "write_off_account", __("Write-off Account"), false, "Account", {
			get_query: () => ({ filters: { company: this.company.get_value(), is_group: 0 } }),
			onchange: () => this.update_balance_state(),
		});
		this.file = this.control($row, "Attach", "payment_file", __("CSV / XLSX File"), true, null, {
			onchange: () => this.preview(),
		});
		this.create_button = this.page.add_inner_button(__("Create Journal Entries"), () => this.confirm_create());
		this.create_button.addClass("btn-primary").prop("disabled", true);
	}

	control(parent, fieldtype, fieldname, label, reqd, options, extra = {}) {
		const $column = $('<div class="col-sm-6 col-lg-4"></div>').appendTo(parent);
		return frappe.ui.form.make_control({
			parent: $column,
			df: { fieldtype, fieldname, label, reqd: reqd ? 1 : 0, options, ...extra },
			render_input: true,
		});
	}

	async preview() {
		if (!this.file.get_value()) return;
		frappe.dom.freeze(__("Checking invoices and outstanding balances..."));
		try {
			const { message } = await frappe.call({
				method: "erpnext_payment_uploading.erpnext_payment_uploading.page.payment_upload.payment_upload.preview",
				args: { file_url: this.file.get_value() },
			});
			this.rows = message.rows;
			this.sync_debit_rows();
			this.render(message.invalid_rows);
			this.update_balance_state(message.invalid_rows);
		} finally {
			frappe.dom.unfreeze();
		}
	}

	render(invalid) {
		const valid = this.rows.length - invalid;
		this.$body.find(".summary").html(
			`<span class="indicator blue">${this.rows.length} ${__("Rows")}</span>
			 <span class="indicator green">${valid} ${__("Ready")}</span>
			 <span class="indicator red">${invalid} ${__("Blocked")}</span>`
		);
		const esc = frappe.utils.escape_html;
		const rows = this.rows.map((r) => `<tr class="${r.status === "Invalid" ? "text-danger" : ""}" title="${esc(r.message || "")}">
			<td>${esc(r.customer || "")}</td><td>${esc(r.invoice_no || "")}</td><td>${esc(r.cheque_no || "")}</td>
			<td class="text-right">${format_currency(r.invoice_amount, r.currency)}</td>
			<td class="text-right">${format_currency(r.ewt_amount, r.currency)}</td>
			<td class="text-right">${format_currency(r.outstanding, r.currency)}</td>
			<td class="text-right">${format_currency(r.write_off_amount, r.currency)}</td></tr>`).join("");
		const issues = this.rows.filter((r) => r.message).map((r) =>
			`<li class="${r.status === "Invalid" ? "text-danger" : "text-warning"}">${esc(r.invoice_no)}: ${esc(r.message)}</li>`
		).join("");
		this.$body.find(".preview").html(`<div class="table-responsive"><table class="table table-bordered table-hover">
			<thead><tr><th>${__("Customer")}</th><th>${__("Sales Invoice No")}</th><th>${__("Customer Ref No")}</th>
			<th>${__("Paid Amount")}</th><th>${__("EWT")}</th><th>${__("Outstanding")}</th><th>${__("Write-off")}</th>
			</tr></thead><tbody>${rows}</tbody>
		</table></div>${issues ? `<div class="mt-2"><strong>${__("Validation Notes")}</strong><ul>${issues}</ul></div>` : ""}`);
		this.render_debits();
	}

	sync_debit_rows() {
		const cheques = [...new Set(this.rows.map((row) => row.cheque_no))];
		this.debit_rows = this.debit_rows.filter((row) => cheques.includes(row.cheque_no));
		for (const cheque_no of cheques) {
			if (!this.debit_rows.some((row) => row.cheque_no === cheque_no)) {
				const cheque_rows = this.rows.filter((row) => row.cheque_no === cheque_no);
				const supplied = cheque_rows.reduce((total, row) => total + flt(row.invoice_amount), 0);
				this.debit_rows.push({ cheque_no, amount: supplied, role: __("Column E Debit"), controls: null });
			}
		}
	}

	add_debit(cheque_no) {
		this.debit_rows.push({ cheque_no, amount: 0, controls: null });
		this.render_debits();
	}

	render_debits() {
		const $section = this.$body.find(".debit-section").empty();
		if (!this.rows.length) return;
		$section.append(`<div class="d-flex justify-content-between align-items-center mb-2">
			<h4 class="m-0">${__("Debit Entries and Balance Check")}</h4>
			<span class="text-muted">${__("Add bank, EWT, receivable, or other debit lines per cheque.")}</span>
		</div>`);
		for (const cheque_no of [...new Set(this.rows.map((row) => row.cheque_no))]) {
			const $card = $(`<div class="frappe-card p-3 mb-3">
				<div class="d-flex justify-content-between"><strong>${__("Cheque")} ${frappe.utils.escape_html(cheque_no)}</strong>
				<button class="btn btn-xs btn-default add-debit">${__("Add Debit")}</button></div>
				<div class="debit-rows mt-2"></div><div class="balance mt-2"></div>
			</div>`).appendTo($section);
			$card.find(".add-debit").on("click", () => this.add_debit(cheque_no));
			for (const row of this.debit_rows.filter((value) => value.cheque_no === cheque_no)) {
				this.render_debit_row($card.find(".debit-rows"), row);
			}
		}
		this.update_balance_state();
	}

	render_debit_row($parent, row) {
		const $row = $('<div class="row align-items-end border-top pt-2 mb-2"></div>').appendTo($parent);
		$(`<div class="col-12 text-muted small mb-1">${frappe.utils.escape_html(row.role || __("Additional Debit"))}</div>`).appendTo($row);
		const make = (fieldtype, fieldname, label, options, width = "col-lg-2") => {
			const $cell = $(`<div class="col-sm-6 ${width}"></div>`).appendTo($row);
			return frappe.ui.form.make_control({ parent: $cell, df: { fieldtype, fieldname, label, options }, render_input: true });
		};
		const account = make("Link", "account", __("Account"), "Account", "col-lg-3");
		account.get_query = () => ({ filters: { company: this.company.get_value(), is_group: 0 } });
		const party_type = make("Select", "party_type", __("Party Type"), "\nCustomer\nSupplier\nEmployee");
		const party = make("Link", "party", __("Party"), row.party_type || "Customer", "col-lg-3");
		const amount = make("Currency", "amount", __("Debit Amount"), null, "col-lg-2");
		const $remove = $('<div class="col-lg-2"><button class="btn btn-xs btn-danger mb-2">Remove</button></div>').appendTo($row);

		account.set_value(row.account || "");
		party_type.set_value(row.party_type || "");
		party.set_value(row.party || "");
		amount.set_value(row.amount || 0);
		row.controls = { account, party_type, party, amount };
		const changed = () => this.update_balance_state();
		account.df.onchange = changed;
		amount.df.onchange = changed;
		party_type.df.onchange = () => {
			party.df.options = party_type.get_value() || "Customer";
			party.set_value("");
			party.refresh();
			changed();
		};
		party.df.onchange = changed;
		$remove.find("button").on("click", () => {
			this.debit_rows = this.debit_rows.filter((value) => value !== row);
			this.render_debits();
		});
	}

	get_debits() {
		return this.debit_rows.map((row) => ({
			cheque_no: row.cheque_no,
			account: row.controls?.account.get_value() || row.account || "",
			party_type: row.controls?.party_type.get_value() || row.party_type || "",
			party: row.controls?.party.get_value() || row.party || "",
			amount: flt(row.controls?.amount.get_value() ?? row.amount),
			remark: row.role || "",
		}));
	}

	update_balance_state(invalid_rows) {
		const invalid = invalid_rows ?? this.rows.filter((row) => row.status === "Invalid").length;
		let balanced = Boolean(this.rows.length);
		for (const cheque_no of [...new Set(this.rows.map((row) => row.cheque_no))]) {
			const invoice_credits = new Map();
			this.rows.filter((row) => row.cheque_no === cheque_no)
				.forEach((row) => invoice_credits.set(row.invoice_no, flt(row.outstanding)));
			const credits = [...invoice_credits.values()].reduce((sum, amount) => sum + amount, 0);
			const paid_debits = this.get_debits().filter((row) => row.cheque_no === cheque_no).reduce((sum, row) => sum + flt(row.amount), 0);
			const ewt = Math.round((this.rows.filter((row) => row.cheque_no === cheque_no)
				.reduce((sum, row) => sum + flt(row.ewt_amount), 0) + Number.EPSILON) * 100) / 100;
			const debits = paid_debits + ewt;
			const difference = debits - credits;
			const can_write_off = Math.abs(difference) < 0.005 || Boolean(this.write_off.get_value());
			const has_ewt_account = !ewt || Boolean(this.ewt_account.get_value());
			const complete = this.get_debits().filter((row) => row.cheque_no === cheque_no)
				.every((row) => row.account && row.amount > 0 && ((!row.party_type && !row.party) || (row.party_type && row.party)));
			balanced = balanced && can_write_off && has_ewt_account && complete;
			this.$body.find(".debit-section .frappe-card").filter((_, card) => $(card).find("strong").text().endsWith(cheque_no))
				.find(".balance").html(`<span class="indicator ${can_write_off ? "green" : "orange"}">
				${__("Paid Debits")}: ${format_currency(paid_debits)} · ${__("EWT")}: ${format_currency(ewt)} ·
				${__("Invoice Credits")}: ${format_currency(credits)} ·
				${__("Write-off")}: ${format_currency(Math.abs(difference))} ${difference >= 0 ? __("Credit") : __("Debit")}</span>`);
		}
		this.create_button.prop("disabled", invalid > 0 || !balanced);
	}

	confirm_create() {
		for (const field of [this.company, this.posting_date, this.ewt_account, this.file]) {
			if (!field.get_value()) {
				frappe.msgprint(__("Please complete all required fields."));
				return;
			}
		}
		frappe.confirm(__("Create balanced draft Journal Entries and open the first entry?"), async () => {
			const { message } = await frappe.call({
				method: "erpnext_payment_uploading.erpnext_payment_uploading.page.payment_upload.payment_upload.create_journal_entries",
				args: {
					rows: this.rows, debits: this.get_debits(), company: this.company.get_value(),
					posting_date: this.posting_date.get_value(), ewt_account: this.ewt_account.get_value(),
					write_off_account: this.write_off.get_value(),
				},
				freeze: true, freeze_message: __("Creating draft Journal Entries..."),
			});
			frappe.msgprint(__("Created {0} draft Journal Entries: {1}", [message.count, message.journal_entries.join(", ")]));
			if (message.journal_entries.length) frappe.set_route("Form", "Journal Entry", message.journal_entries[0]);
		});
	}
}

File: erpnext_payment_uploading/erpnext_payment_uploading/page/payment_upload/payment_upload.py
import csv
import io
from collections import OrderedDict
from decimal import Decimal, InvalidOperation, ROUND_HALF_UP

import frappe
from frappe import _
from frappe.utils import flt, getdate, nowdate


@frappe.whitelist()
def preview(file_url):
    _require_create_permission()
    rows = _read_rows(file_url)
    if not rows:
        frappe.throw(_("The uploaded file has no payment rows."))

    names = list({row.invoice_no for row in rows if row.invoice_no})
    invoices = {
        row.name: row
        for row in frappe.get_all(
            "Sales Invoice",
            filters={"name": ["in", names]},
            fields=["name", "customer", "customer_name", "outstanding_amount", "currency", "docstatus", "debit_to"],
            limit_page_length=0,
        )
    }
    total_ewt = _money(sum((Decimal(str(row.ewt_amount)) for row in rows), Decimal("0")))
    allocated_ewt = Decimal("0")
    result = []
    for index, row in enumerate(rows):
        row_ewt = total_ewt - allocated_ewt if index == len(rows) - 1 else _money(row.ewt_amount)
        allocated_ewt += row_ewt
        invoice = invoices.get(row.invoice_no)
        errors, warnings = [], []
        if not row.cheque_no:
            errors.append(_("Check # is required"))
        if not row.invoice_no:
            errors.append(_("Invoice # is required"))
        elif not invoice:
            errors.append(_("Sales Invoice was not found"))
        elif invoice.docstatus != 1:
            errors.append(_("Sales Invoice is not submitted"))
        elif flt(invoice.outstanding_amount) <= 0:
            errors.append(_("Sales Invoice has no outstanding balance"))
        if abs(row.basis_amount - row.ewt_amount - row.cheque_amount) > Decimal("0.02"):
            errors.append(_("Calculated net amount does not equal basis less EWT"))
        if invoice and abs(row.invoice_amount - Decimal(str(invoice.outstanding_amount or 0))) > Decimal("0.02"):
            warnings.append(_("Uploaded amount differs from current outstanding"))

        result.append(
            {
                **row,
                "invoice_amount": float(row.invoice_amount),
                "basis_amount": float(row.basis_amount),
                "ewt_amount": float(row_ewt),
                "cheque_amount": float(row.cheque_amount),
                "difference_amount": float(row.difference_amount),
                "customer": invoice.customer if invoice else None,
                "customer_name": invoice.customer_name if invoice else None,
                "currency": invoice.currency if invoice else None,
                "outstanding": flt(invoice.outstanding_amount) if invoice else 0,
                "write_off_amount": float(
                    _money(invoice.outstanding_amount) - _money(row.invoice_amount) - row_ewt
                )
                if invoice
                else 0,
                "status": "Invalid" if errors else ("Warning" if warnings else "Valid"),
                "message": "; ".join(errors + warnings),
            }
        )
    result.sort(key=lambda row: (row.get("customer") or "", row.get("cheque_no") or "", row.get("row_no") or 0))
    return {"rows": result, "invalid_rows": sum(r["status"] == "Invalid" for r in result)}


@frappe.whitelist()
def create_journal_entries(
    rows,
    debits,
    company,
    posting_date,
    check_amount_received,
    ewt_account=None,
    write_off_account=None,
):
    _require_create_permission()
    rows = frappe.parse_json(rows) if isinstance(rows, str) else rows
    debits = frappe.parse_json(debits) if isinstance(debits, str) else debits
    if not rows or not debits or not company or not posting_date or not check_amount_received:
        frappe.throw(_("Rows, debit entries, Company, Posting Date, and Check Amount Received are required."))

    input_rows = [frappe._dict(value) for value in rows]
    uploaded_paid_total = _money(sum((Decimal(str(row.invoice_amount)) for row in input_rows), Decimal("0")))
    if uploaded_paid_total != _money(check_amount_received):
        frappe.throw(
            _("Check/Bank/Debit Amount ({0}) must equal uploaded Paid Amount total ({1}).").format(
                _money(check_amount_received), uploaded_paid_total
            )
        )
    invoice_names = list({row.invoice_no for row in input_rows})
    all_invoices = {
        row.name: row
        for row in frappe.get_all(
            "Sales Invoice",
            filters={"name": ["in", invoice_names], "docstatus": 1, "company": company},
            fields=["name", "customer", "outstanding_amount", "currency", "debit_to", "due_date"],
            limit_page_length=0,
        )
    }
    if len(all_invoices) != len(invoice_names):
        frappe.throw(_("One or more invoices are missing, not submitted, or belong to another company."))

    company_currency = frappe.get_cached_value("Company", company, "default_currency")
    invoice_currencies = {invoice.currency for invoice in all_invoices.values()}
    if invoice_currencies != {company_currency}:
        frappe.throw(
            _("This uploader currently supports company-currency invoices only ({0}). Found: {1}").format(
                company_currency, ", ".join(sorted(invoice_currencies))
            )
        )

    account_names = {row.get("account") for row in debits if row.get("account")}
    account_names.update(invoice.debit_to for invoice in all_invoices.values())
    if ewt_account:
        account_names.add(ewt_account)
    if write_off_account:
        account_names.add(write_off_account)
    account_currencies = {
        row.name: row.account_currency or company_currency
        for row in frappe.get_all(
            "Account",
            filters={"name": ["in", list(account_names)], "company": company, "is_group": 0},
            fields=["name", "account_currency"],
            limit_page_length=0,
        )
    }
    if len(account_currencies) != len(account_names):
        frappe.throw(_("One or more selected accounts are invalid for company {0}.").format(company))
    foreign_accounts = [name for name, currency in account_currencies.items() if currency != company_currency]
    if foreign_accounts:
        frappe.throw(
            _("This uploader currently supports company-currency accounts only. Check: {0}").format(
                ", ".join(foreign_accounts)
            )
        )

    # One uploaded workbook creates one Journal Entry. Customer Ref No remains
    # an invoice-level audit value in User Remark, not a grouping key.
    input_rows.sort(key=lambda row: row.row_no or 0)
    credit_by_invoice = OrderedDict()
    for row in input_rows:
        if row.invoice_no not in credit_by_invoice:
            invoice = all_invoices[row.invoice_no]
            outstanding = _money(invoice.outstanding_amount)
            if outstanding <= 0:
                frappe.throw(_("Invoice {0} has no outstanding balance.").format(row.invoice_no))
            credit_by_invoice[row.invoice_no] = {
                "outstanding": outstanding,
                "paid": Decimal("0"),
                "ewt_raw": Decimal("0"),
            }
        credit_by_invoice[row.invoice_no]["paid"] += Decimal(str(row.invoice_amount))
        credit_by_invoice[row.invoice_no]["ewt_raw"] += Decimal(str(row.ewt_amount))

    debit_rows = [frappe._dict(row) for row in debits]
    for debit in debit_rows:
        if not debit.account or _money(debit.amount) <= 0:
            frappe.throw(_("Every debit row requires an account and positive amount."))
    paid_debit_total = sum((_money(row.amount) for row in debit_rows), Decimal("0"))
    ewt_total = _money(sum((Decimal(str(row.ewt_amount)) for row in input_rows), Decimal("0")))
    if ewt_total and not ewt_account:
        frappe.throw(_("Select an EWT Account."))

    # Required order: configured debit(s), invoice references, write-off, EWT.
    accounts = []
    for debit in debit_rows:
        account_row = {
            "account": debit.account,
            "account_currency": company_currency,
            "exchange_rate": 1,
            "debit_in_account_currency": float(_money(debit.amount)),
            "user_remark": debit.get("remark") or _("Check/Bank/Debit Amount"),
        }
        if debit.get("party_type") and debit.get("party"):
            account_row.update({"party_type": debit.party_type, "party": debit.party})
        accounts.append(account_row)

    invoice_names_in_order = list(credit_by_invoice)
    allocated_ewt = Decimal("0")
    for index, name in enumerate(invoice_names_in_order):
        values = credit_by_invoice[name]
        values["ewt"] = (
            ewt_total - allocated_ewt
            if index == len(invoice_names_in_order) - 1
            else _money(values["ewt_raw"])
        )
        allocated_ewt += values["ewt"]

    for name, values in credit_by_invoice.items():
        invoice = all_invoices[name]
        accounts.append(
            {
                "account": invoice.debit_to,
                "account_currency": company_currency,
                "exchange_rate": 1,
                "party_type": "Customer",
                "party": invoice.customer,
                "credit_in_account_currency": float(values["outstanding"]),
                "reference_type": "Sales Invoice",
                "reference_name": name,
                "reference_due_date": invoice.due_date,
                "user_remark": name,
            }
        )

    # Write-off is calculated and tagged per invoice. Multiple invoice
    # differences therefore produce multiple Journal Entry Account rows.
    for name, values in credit_by_invoice.items():
        invoice = all_invoices[name]
        difference = _money(values["paid"]) + values["ewt"] - values["outstanding"]
        if abs(difference) <= Decimal("0.005"):
            continue
        write_off_row = {
            "account": write_off_account,
            "account_currency": company_currency,
            "exchange_rate": 1,
            "party_type": "Customer",
            "party": invoice.customer,
            "reference_type": "Sales Invoice",
            "reference_name": name,
            "reference_due_date": invoice.due_date,
            "user_remark": _("Write-off for {0}").format(name),
        }
        if difference > 0:
            write_off_row["credit_in_account_currency"] = float(difference)
        else:
            write_off_row["debit_in_account_currency"] = float(abs(difference))
        accounts.append(write_off_row)

    if ewt_total:
        accounts.append(
            {
                "account": ewt_account,
                "account_currency": company_currency,
                "exchange_rate": 1,
                "debit_in_account_currency": float(ewt_total),
                "user_remark": _("EWT"),
            }
        )

    customer_refs = list(dict.fromkeys(_clean_id(row.cheque_no) for row in input_rows))
    first = input_rows[0]
    remark_lines = [
            "\t".join(
                [
                    _clean_id(row.cheque_no),
                    _display_date(posting_date),
                    _display_date(row.cheque_date),
                    row.invoice_no,
                    "",
                    "{:,.2f}".format(Decimal(str(row.invoice_amount))),
                ]
            )
            for row in input_rows
        ]
    doc = frappe.get_doc(
            {
                "doctype": "Journal Entry",
                "voucher_type": "Bank Entry",
                "company": company,
                "posting_date": getdate(posting_date),
                "cheque_no": ", ".join(customer_refs)[:140],
                "cheque_date": getdate(first.cheque_date or nowdate()),
                "user_remark": "\n".join(remark_lines),
                "accounts": accounts,
            }
        )
    doc.insert()
    return {"journal_entries": [doc.name], "count": 1}


def _read_rows(file_url):
    file_doc = frappe.get_doc("File", {"file_url": file_url})
    content = file_doc.get_content()
    filename = (file_doc.file_name or file_url).lower()
    if filename.endswith(".csv"):
        text = content.decode("utf-8-sig") if isinstance(content, bytes) else content
        raw = list(csv.DictReader(io.StringIO(text)))
    elif filename.endswith(".xlsx"):
        from openpyxl import load_workbook

        workbook = load_workbook(io.BytesIO(content), read_only=True, data_only=True)
        if "Sheet1" in workbook.sheetnames:
            sheet1 = workbook["Sheet1"]
            if _header(sheet1.cell(row=3, column=4).value) in {"si", "invoice", "invoice #"}:
                return _read_sheet1_columns(sheet1)
        sheet = workbook["july 24 2026"] if "july 24 2026" in workbook.sheetnames else workbook.active
        values = list(sheet.iter_rows(values_only=True))
        header_index = next((i for i, values_row in enumerate(values) if "invoice #" in {_header(v) for v in values_row}), None)
        if header_index is None:
            frappe.throw(_("Could not find a row containing the Invoice # header."))
        headers = [str(v or "") for v in values[header_index]]
        raw = [dict(zip(headers, values_row)) for values_row in values[header_index + 1 :] if any(v not in (None, "") for v in values_row)]
    else:
        frappe.throw(_("Please upload a CSV or XLSX file."))

    rows = []
    for index, source in enumerate(raw, start=2):
        value = {_header(k): v for k, v in source.items()}
        try:
            invoice_amount = _decimal(value.get("invoice amount"))
            ewt_amount = _decimal(value.get("ewt1%") or value.get("ewt 1%"))
            cheque_amount = _decimal(value.get("check amount") or value.get("cheque amount"))
        except InvalidOperation:
            frappe.throw(_("Row {0} contains an invalid monetary amount.").format(index))
        rows.append(
            frappe._dict(
                row_no=index,
                cheque_no=_clean_id(value.get("check #") or value.get("cheque #")),
                cheque_date=str(getdate(value.get("check date") or value.get("cheque date"))) if value.get("check date") or value.get("cheque date") else None,
                invoice_no=_clean_id(value.get("invoice #")),
                invoice_amount=invoice_amount,
                basis_amount=invoice_amount,
                ewt_amount=ewt_amount,
                cheque_amount=cheque_amount,
                difference_amount=invoice_amount - cheque_amount,
            )
        )
    return rows


def _read_sheet1_columns(sheet):
    """Read Sheet1 where only B (cheque), D (invoice), and E (amount) are supplied."""
    rows = []
    cheque_date = sheet.cell(row=2, column=2).value
    # Streaming is essential here. Random-access ``sheet.cell`` calls on a
    # read-only worksheet repeatedly scan its XML and become extremely slow
    # for production files with more than a thousand rows.
    for index, values in enumerate(
        sheet.iter_rows(min_row=4, max_col=5, values_only=True), start=4
    ):
        cheque_no = values[1]
        invoice_no = values[3]
        source_amount = values[4]
        if cheque_no in (None, "") and invoice_no in (None, "") and source_amount in (None, ""):
            continue
        if cheque_no in (None, "") or invoice_no in (None, "") or source_amount in (None, ""):
            frappe.throw(_("Sheet1 row {0} must contain columns B, D, and E.").format(index))
        try:
            invoice_amount = _decimal(source_amount)
        except InvalidOperation:
            frappe.throw(_("Sheet1 row {0} contains an invalid amount in column E.").format(index))
        basis_amount = invoice_amount * Decimal("1.01")
        ewt_amount = basis_amount * Decimal("0.01")
        cheque_amount = basis_amount - ewt_amount
        difference_amount = invoice_amount - cheque_amount
        rows.append(
            frappe._dict(
                row_no=index,
                cheque_no=_clean_id(cheque_no),
                cheque_date=str(getdate(cheque_date)) if cheque_date else None,
                invoice_no=_clean_id(invoice_no),
                invoice_amount=invoice_amount,
                basis_amount=basis_amount,
                ewt_amount=ewt_amount,
                cheque_amount=cheque_amount,
                difference_amount=difference_amount,
            )
        )
    return rows


def _header(value):
    return " ".join(str(value or "").strip().lower().split())


def _clean_id(value):
    if value is None:
        return ""
    if isinstance(value, float) and value.is_integer():
        value = int(value)
    return str(value).replace("\u00a0", " ").strip()


def _decimal(value):
    return Decimal(str(value or 0).replace(",", ""))


def _money(value):
    return _decimal(value).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)


def _display_date(value):
    value = getdate(value)
    return "{0}/{1}/{2}".format(value.month, value.day, value.year)


def _require_create_permission():
    if not frappe.has_permission("Journal Entry", "create"):
        frappe.throw(_("You do not have permission to create Journal Entries."), frappe.PermissionError)
